nogo: include and exclude targets in configuration

<!-- Please answer these questions before submitting your issue. Thanks! -->

### What version of rules_go are you using?
```
 http_archive(
        name = "io_bazel_rules_go",
        sha256 = "87f0fb9747854cb76a0a82430adccb6269f7d394237104a4523b51061c469171",
        urls = [
            "https://mirror.bazel.build/github.com/bazelbuild/rules_go/releases/download/v0.23.1/rules_go-v0.23.1.tar.gz",
            "https://github.com/bazelbuild/rules_go/releases/download/v0.23.1/rules_go-v0.23.1.tar.gz",
        ],
    )
```

### What version of gazelle are you using?

```
    http_archive(
        name = "bazel_gazelle",
        sha256 = "bfd86b3cbe855d6c16c6fce60d76bd51f5c8dbc9cfcaef7a2bb5c1aafd0710e8",
        urls = [
            "https://mirror.bazel.build/github.com/bazelbuild/bazel-gazelle/releases/download/v0.21.0/bazel-gazelle-v0.21.0.tar.gz",
            "https://github.com/bazelbuild/bazel-gazelle/releases/download/v0.21.0/bazel-gazelle-v0.21.0.tar.gz",
        ],
    )
```
### What version of Bazel are you using?
`bazel 3.1.0-homebrew`

### Does this issue reproduce with the latest releases of all the above?
Yes


### What operating system and processor architecture are you using?
Macos


### Any other potentially useful information about your toolchain?
I am using go modules. 

### What did you do?
`  go_register_toolchains(nogo = "@io_bazel_rules_go//:tools_nogo")`
I want to use nogo. I followed the instructions ( I know it mentions this) and it lines external packages which I have no control over. Is there a way I can limit it my code ? 

### What did you expect to see?
Lint only my code 

### What did you see instead?
Lints everything in the external. 
```
external/org_golang_x_tools/go/vcs/vcs.go:532:2: declaration of "metaImport" shadows declaration at line 599
```

--- a/go/nogo.rst
+++ b/go/nogo.rst
@@ -204,6 +204,9 @@ contain the following key-value pairs:
 | ``flag.Value`` specified by the analyzer.                                                        |
 +----------------------------+---------------------------------------------------------------------+
 
+``nogo`` also supports a special key to specify the same config for all analyzers, even if they are
+not explicitly specified called ``_base``. See below for an example of its usage.
+
 Example
 ^^^^^^^
 
@@ -216,6 +219,12 @@ on a command line driver.
 .. code:: json
 
     {
+      "_base": {
+        "description": "Base config that all subsequent analyzers, even unspecified will inherit.",
+        "exclude_files": {
+          "third_party/": "exclude all third_party code for all analyzers"
+        }
+      },
       "importunsafe": {
         "exclude_files": {
           "src/foo\\.go": "manually verified that behavior is working-as-intended",

--- a/go/tools/builders/nogo_main.go
+++ b/go/tools/builders/nogo_main.go
@@ -43,6 +43,8 @@ import (
 	"golang.org/x/tools/go/gcexportdata"
 )
 
+const nogoBaseConfigName = "_base"
+
 func init() {
 	if err := analysis.Validate(analyzers); err != nil {
 		log.Fatal(err)
@@ -89,7 +91,7 @@ func run(args []string) error {
 		return fmt.Errorf("errors found by nogo during build-time code analysis:\n%s\n", diagnostics)
 	}
 	if *xPath != "" {
-		if err := ioutil.WriteFile(abs(*xPath), facts, 0666); err != nil {
+		if err := ioutil.WriteFile(abs(*xPath), facts, 0o666); err != nil {
 			return fmt.Errorf("error writing facts: %v", err)
 		}
 	}
@@ -399,10 +401,26 @@ func checkAnalysisResults(actions []*action, pkg *goPackage) string {
 		if len(act.diagnostics) == 0 {
 			continue
 		}
-		config, ok := configs[act.a.Name]
-		if !ok {
-			// If the analyzer is not explicitly configured, it emits diagnostics for
-			// all files.
+		var currentConfig config
+		// Use the base config if it exists.
+		if baseConfig, ok := configs[nogoBaseConfigName]; ok {
+			currentConfig = baseConfig
+		}
+		// Overwrite the config with the desired config. Any unset fields
+		// in the config will default to the base config.
+		if actionConfig, ok := configs[act.a.Name]; ok {
+			if actionConfig.analyzerFlags != nil {
+				currentConfig.analyzerFlags = actionConfig.analyzerFlags
+			}
+			if actionConfig.onlyFiles != nil {
+				currentConfig.onlyFiles = actionConfig.onlyFiles
+			}
+			if actionConfig.excludeFiles != nil {
+				currentConfig.excludeFiles = actionConfig.excludeFiles
+			}
+		}
+
+		if currentConfig.onlyFiles == nil && currentConfig.excludeFiles == nil {
 			for _, diag := range act.diagnostics {
 				diagnostics = append(diagnostics, entry{Diagnostic: diag, Analyzer: act.a})
 			}
@@ -418,10 +436,10 @@ func checkAnalysisResults(actions []*action, pkg *goPackage) string {
 				filename = p.Filename
 			}
 			include := true
-			if len(config.onlyFiles) > 0 {
+			if len(currentConfig.onlyFiles) > 0 {
 				// This analyzer emits diagnostics for only a set of files.
 				include = false
-				for _, pattern := range config.onlyFiles {
+				for _, pattern := range currentConfig.onlyFiles {
 					if pattern.MatchString(filename) {
 						include = true
 						break
@@ -429,7 +447,7 @@ func checkAnalysisResults(actions []*action, pkg *goPackage) string {
 				}
 			}
 			if include {
-				for _, pattern := range config.excludeFiles {
+				for _, pattern := range currentConfig.excludeFiles {
 					if pattern.MatchString(filename) {
 						include = false
 						break

--- a/tests/core/nogo/custom/custom_test.go
+++ b/tests/core/nogo/custom/custom_test.go
@@ -297,6 +297,24 @@ func run(pass *analysis.Pass) (interface{}, error) {
   }
 }
 
+-- baseconfig.json --
+{
+  "_base": {
+    "exclude_files": {
+      "has_.*\\.go": "Visibility analyzer not specified. Still inherits this special exception."
+    }
+  },
+  "importfmt": {
+    "only_files": {
+      "has_errors\\.go": ""
+    }
+  },
+  "foofuncname": {
+    "description": "no exemptions since we know this check is 100% accurate, so override base config",
+    "exclude_files": {}
+  }
+}
+
 -- has_errors.go --
 package haserrors
 
@@ -409,6 +427,18 @@ func Test(t *testing.T) {
 		}, {
 			desc:        "custom_config_linedirective",
 			config:      "config.json",
+			target:      "//:has_errors_linedirective",
+			wantSuccess: false,
+			includes: []string{
+				`linedirective_foofuncname.go:.*function must not be named Foo \(foofuncname\)`,
+				`linedirective_visibility.go:.*function D is not visible in this package \(visibility\)`,
+			},
+			excludes: []string{
+				`importfmt`,
+			},
+		}, {
+			desc:        "custom_config_with_base_linedirective",
+			config:      "baseconfig.json",
 			target:      "//:has_errors_linedirective",
 			wantSuccess: false,
 			includes: []string{

